extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -0,0 +1,526 @@
+  private final InputRowParser<ByteBuffer> parser;
+  private final KafkaTuningConfig tuningConfig;
+  private final KafkaIOConfig ioConfig;
+
+  private volatile Appenderator appenderator = null;
+  private volatile FireDepartmentMetrics fireDepartmentMetrics = null;
+  private volatile boolean stopping = false;
+  private volatile boolean publishing = false;
+  private volatile Thread runThread = null;
+
+  @JsonCreator
+  public KafkaIndexTask(
+      @JsonProperty("id") String id,
+      @JsonProperty("resource") TaskResource taskResource,
+      @JsonProperty("dataSchema") DataSchema dataSchema,
+      @JsonProperty("tuningConfig") KafkaTuningConfig tuningConfig,
+      @JsonProperty("ioConfig") KafkaIOConfig ioConfig,
+      @JsonProperty("context") Map<String, Object> context
+  )
+  {
+    super(
+        id == null ? makeTaskId(dataSchema.getDataSource(), RANDOM.nextInt()) : id,
+        String.format("%s_%s", TYPE, dataSchema.getDataSource()),
+        taskResource,
+        dataSchema.getDataSource(),
+        context
+    );
+
+    this.dataSchema = Preconditions.checkNotNull(dataSchema, "dataSchema");
+    this.parser = Preconditions.checkNotNull((InputRowParser<ByteBuffer>) dataSchema.getParser(), "parser");
+    this.tuningConfig = Preconditions.checkNotNull(tuningConfig, "tuningConfig");
+    this.ioConfig = Preconditions.checkNotNull(ioConfig, "ioConfig");
+  }
+
+  private static String makeTaskId(String dataSource, int randomBits)
+  {
+    final StringBuilder suffix = new StringBuilder(8);
+    for (int i = 0; i < Ints.BYTES * 2; ++i) {
+      suffix.append((char) ('a' + ((randomBits >>> (i * 4)) & 0x0F)));
+    }
+    return Joiner.on("_").join(TYPE, dataSource, suffix);
+  }
+
+  @Override
+  public String getType()
+  {
+    return TYPE;
+  }
+
+  @Override
+  public boolean isReady(TaskActionClient taskActionClient) throws Exception
+  {
+    return true;
+  }
+
+  @JsonProperty
+  public DataSchema getDataSchema()
+  {
+    return dataSchema;
+  }
+
+  @JsonProperty
+  public KafkaTuningConfig getTuningConfig()
+  {
+    return tuningConfig;
+  }
+
+  @JsonProperty
+  public KafkaIOConfig getIoConfig()
+  {
+    return ioConfig;
+  }
+
+  @Override
+  public TaskStatus run(final TaskToolbox toolbox) throws Exception
+  {
+    log.info("Starting up!");
+
+    runThread = Thread.currentThread();
+
+    // Set up FireDepartmentMetrics
+    final FireDepartment fireDepartmentForMetrics = new FireDepartment(
+        dataSchema,
+        new RealtimeIOConfig(null, null, null),
+        null
+    );
+    fireDepartmentMetrics = fireDepartmentForMetrics.getMetrics();
+    toolbox.getMonitorScheduler().addMonitor(
+        new RealtimeMetricsMonitor(
+            ImmutableList.of(fireDepartmentForMetrics),
+            ImmutableMap.of(DruidMetrics.TASK_ID, new String[]{getId()})
+        )
+    );
+
+    try (
+        final Appenderator appenderator0 = newAppenderator(fireDepartmentMetrics, toolbox);
```

Comment:
they're the same appenderator, just some annoying try-with-resources stuff going on here.